File: apps/web/src/lib/api/senderAttestation.ts
"use client";

/**
 * The outbound-sender confirmation a client records against one of their own numbers
 * (`apps/api/campaigns/sender_attestation.py`).
 *
 * TRAI direction RG-25/(18)/2023-QoS (E-10291), 18 Jun 2024 forbids a sender from placing
 * promotional, service or transactional voice calls from an ordinary 10-digit number, and
 * names the delegation chain — employees, channel partners, DSAs, BPO partner, outsourced
 * call centre — so the obligation cannot be handed to a vendor
 * (`docs/evidence/primary-legal-findings-2026-09-20.md` §1). Under Model B (D-474) the
 * client holds the carrier account and IS that sender, which is why the only write path is
 * a client-realm one: there is no admin route that confirms on their behalf.
 *
 * Three properties of the API this module keeps rather than smooths over:
 *
 * - **`applicable` is the SERVER's predicate and is never re-derived here.** It is false
 *   for a registered 140 or 160 header, and the POST refuses one
 *   (`sender_attestation_not_applicable`). A screen that decided for itself which numbers
 *   need the confirmation would eventually offer a control whose only outcome is a refusal.
 * - **`attested` is false when the stored row predates the current wording.** The client
 *   agreed to different words, so the answer is "ask again", not "error".
 * - **The withdrawal is a new row, never a deletion** (hard rule 4). `DELETE` is the verb
 *   the route spells; what it records is a change of mind, and no copy built on this may
 *   say the earlier confirmation is gone.
 */

import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from "@tanstack/react-query";

import { apiRequest, type Session } from "./client";
import type { components } from "./schema";

export type SenderAttestation = components["schemas"]["SenderAttestationOut"];

/** The refusal when the wording moved while the page was open. The remedy is a reload. */
export const STALE_STATEMENT_CODE = "sender_statement_not_current";

/** The act `rbac.VIEW_AS_WITHHELD_ACTS` withholds from a view-as operator. */
export const SENDER_ATTESTATION_ACT = "compliance.outbound_sender_attestation";

function path(numberId: string): string {
  return `/v1/numbers/${numberId}/sender-attestation`;
}

export function senderAttestationKey(org: string, numberId: string): readonly string[] {
  return ["sender-attestation", org, numberId] as const;
}

/** What this number's confirmation says today, and whether it is even needed. */
export function useSenderAttestation(
  session: Session,
  numberId: string,
): UseQueryResult<SenderAttestation> {
  return useQuery({
    queryKey: senderAttestationKey(session.orgSlug, numberId),
    queryFn: () => apiRequest<SenderAttestation>(session, path(numberId)),
  });
}

/**
 * Both writes, as one hook, because they are one decision in two directions and their
 * cache consequences are identical.
 *
 * The response of each is written into the query's own cache rather than only
 * invalidated: the panel's next paint is the answer to "did that land", and a refetch
 * round trip would show the previous state in between.
 *
 * The launch gate reads this fact too (`number_series_mismatch`), and a client who
 * confirms on this screen and then finds their campaign still refused would reasonably
 * conclude the confirmation did nothing — so every campaign's check is invalidated with
 * it. Broad rather than keyed by campaign: this number may be attached to any of them,
 * and this module cannot know which.
 */
export function useSetSenderAttestation(
  session: Session,
  numberId: string,
): UseMutationResult<SenderAttestation, Error, { withdraw: boolean; statementVersion: string }> {
  const client = useQueryClient();
  return useMutation({
    mutationFn: ({ withdraw, statementVersion }) =>
      withdraw
        ? apiRequest<SenderAttestation>(session, path(numberId), { method: "DELETE" })
        : apiRequest<SenderAttestation>(session, path(numberId), {
            method: "POST",
            body: { statement_version: statementVersion },
          }),
    onSuccess: (state) => {
      client.setQueryData(senderAttestationKey(session.orgSlug, numberId), state);
      void client.invalidateQueries({ queryKey: ["campaign-check", session.orgSlug] });
    },
  });
}

File: apps/api/campaigns/sender_attestation.py
"""The client's recorded exception to the ordinary-DID refusal, and the gate that reads it.

TRAI direction RG-25/(18)/2023-QoS (E-10291), 18 Jun 2024: *"Senders shall not use any
other 10-digit fixed line/ mobile number for making Promotional/ Service/ Transactional
voice calls to their customers, either directly or through their employees or channel
partners, DSAs, BPO partner, in-house or outsourced Call Centre, etc."*

The direction binds the SENDER and names the delegation chain — employees, channel
partners, DSAs, BPO partner, outsourced call centre — so the obligation cannot be handed to
a vendor. Under Model B (D-474) the client holds the carrier account and is that sender.

**SO THE REFUSAL IS THE DEFAULT AND THIS IS THE CLIENT'S OWN EXCEPTION TO IT, NOT OURS.**
`SERIES_FOR_CLASSIFICATION` allows only 140 and 160. A `standard` number reaches the dial
path only where a named person at the client has been shown the obligation and accepted it,
and that acceptance is a row with their id and the wording they saw. The alternative
considered and rejected was simply restoring `standard` to the allowed list: that makes the
decision OURS, undocumented, on a surface where the penalty lands on a client who never saw
the rule.

**PROMOTIONAL IS NOT REACHABLE THROUGH THIS AND MUST NOT BECOME SO.** 140 is the only series
that may carry a promotional call. No client declaration makes a promotional call from an
ordinary number lawful, and `attested_series_for` refuses to widen that classification even
when an attestation exists.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final
from uuid import UUID

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from apps.api.core.errors import ProblemError
from apps.api.db.base import uuid7

#: The wording a client accepts, versioned. What the person agreed to is the sentence they
#: were shown, so the row records which one — and changing this string asks every client
#: again rather than letting an old click stand for new words. Same rule
#: `legal.service.record_acceptance` applies to a document version.
SENDER_STATEMENT_VERSION: Final[str] = "2026-09-20"

SENDER_STATEMENT: Final[str] = (
    "I confirm that my business is the sender of these calls, that I have been told TRAI "
    "requires promotional, service and transactional voice calls to be made only from a "
    "registered 140 or 160 series voice header, that this number is not one, and that my "
    "business accepts responsibility for calls made from it."
)

#: The classifications an attestation can open. Promotional is absent deliberately — see
#: this module's docstring.
ATTESTABLE_CLASSIFICATIONS: Final[frozenset[str]] = frozenset({"service", "transactional"})

ATTESTATION_RULE: Final[str] = "outbound_sender_not_attested"


def attestation_reason(series: str, classification: str) -> str:
    """The client-facing refusal. Says what to do, in both of the two ways out."""
    return (
        f"This campaign dials from a {series} number, and TRAI requires {classification} "
        "voice calls to come from a registered 140 or 160 series header. Either attach a "
        "140 or 160 number, or confirm on the number's settings page that your business is "
        "the sender and accepts responsibility for calls from this one."
    )


@dataclass(frozen=True, slots=True)
class AttestationState:
    """The latest row for one number, read as an answer rather than a history."""

    attested: bool
    statement_version: str | None

    @property
    def current(self) -> bool:
        """Attested AND against the wording we show today.

        A stale version is NOT attested for the gate's purposes: the client agreed to
        different words, and treating that as agreement to these would make the row
        evidence of something that did not happen.
        """
        return self.attested and self.statement_version == SENDER_STATEMENT_VERSION


_LATEST_SQL = text(
    "SELECT state, statement_version FROM outbound_sender_attestations "
    "WHERE phone_number_id = :number_id ORDER BY created_at DESC, id DESC LIMIT 1"
)


async def latest_attestation(session: AsyncSession, *, phone_number_id: UUID) -> AttestationState:
    """The current state of one number. Under RLS, so the tenant is implied.

    Ordered by `id` after `created_at` because two rows written inside one transaction share
    `now()`, and a withdrawal that sorted before the attestation it revokes would report the
    opposite of the truth. uuid7 is time-ordered, so the tiebreak is still chronological.
    """
    row = (await session.execute(_LATEST_SQL, {"number_id": phone_number_id})).first()
    if row is None:
        return AttestationState(attested=False, statement_version=None)
    return AttestationState(attested=row[0] == "attested", statement_version=row[1])


async def record_attestation(
    session: AsyncSession,
    *,
    tenant_id: UUID,
    phone_number_id: UUID,
    user_id: UUID,
    statement_version: str,
    withdraw: bool = False,
) -> UUID:
    """Append one attestation or withdrawal. INSERT-only (hard rule 4).

    Refuses a stale statement version before the write, for `record_acceptance`'s reason:
    the wording the person ticked is half the evidence, so a console left open across a
    change must reload rather than have today's version recorded against yesterday's click.
    A WITHDRAWAL is exempt — refusing to let someone retract because their page is old would
    hold them to an obligation they are trying to step out of.
    """
    if not withdraw and statement_version != SENDER_STATEMENT_VERSION:
        raise ProblemError.business_rule(
            "sender_statement_not_current",
            "The confirmation on your screen is out of date.",
            remediation="Reload the page and read the confirmation again before accepting.",
        )
    row_id = uuid7()
    await session.execute(
        text(
            "INSERT INTO outbound_sender_attestations "
            "(id, tenant_id, phone_number_id, state, statement_version, attested_by) "
            "VALUES (:id, :tid, :nid, :state, :sv, :uid)"
        ),
        {
            "id": row_id,
            "tid": tenant_id,
            "nid": phone_number_id,
            "state": "withdrawn" if withdraw else "attested",
            "sv": SENDER_STATEMENT_VERSION,
            "uid": user_id,
        },
    )
    return row_id


__all__ = [
    "ATTESTABLE_CLASSIFICATIONS",
    "ATTESTATION_RULE",
    "SENDER_STATEMENT",
    "SENDER_STATEMENT_VERSION",
    "AttestationState",
    "attestation_reason",
    "latest_attestation",
    "record_attestation",
]
